### ISSUE TITLE
Find locale filename by language codes

### ISSUE THREAD

**Issue by codeofsumit** [moment/moment#3514]
In our build process we're trying to only load the locale we need for the current language to decrease file size.
I managed to do that but, in order to load the correct locale, I need it's name/filename by languagecode.

Currently, I just replace `_` with `-`, put all to lowercase and if country and language are equal, delete the country part.

Examples:

* `en_GB` -> `en-gb`
* `de_DE` -> `de`
* `de_AT` -> `de-at`
* **`en_US` -> `en-us`** -> error

And there's the problem. `en-us.js` is not a filename in locale in moment.js.

So my question is: what is a recommended, reliable way to get the moment.js locale filename by language code, without using moment.js?

**Comment by icambron** [comment]
Internally, Moment does it by trying `en-us-whatever`, `en-us`, `en`, etc them until one succeeds. The [code](https://github.com/moment/moment/blob/develop/src/lib/locale/locales.js#L23) is a little more complicated than it needs to be for your use case because it accepts a whole array of user-requested locales and does the `en-us` -> `en` backoff with lower priority than trying other items in the array. But the principle is the same; there's no way to know whether a locale def will be there until you get the locale def. This makes sense; we have no idea what additional locale defs you might have put on your server or compiled into Moment. So you'd have to do the same.

You could, of course, make a copy of `en` and name it `en-us`, but I don't know if your use case also requires you to handle arbitrary codes, like `de-DE-bavarian`, that Moment can only handle by progressive fallback.

I hope that helps. I'm going to close this, but if you follow-ups, just @ me and I'll continue to respond.

**Comment by codeofsumit** [comment]
@icambron thanks for the info that sounds reasonable. 
TL;DR: Do you know any other momentjs locale filenames where the filename is not part of the language code?

But if `en-us` falls back to `en` I still have to get the filename, which is `en-gb` - as far as i can see, there is no `en.js`.
So even trial and error style of getting those files won't succeed for `en_US` because the filename needed is `en-gb.js`.
I'm ok with making an exception in the code for `en_US`, as long as I know all exceptions that have to be made.
Do you know any other momentjs locale filenames where the filename is not part of the language code?

**Comment by icambron** [comment]
Ah, right, I left that part out: `en` is the default that comes with Moment; `moment.locale('en')` succeeds always. So you want the progressive fallback, but when you reach `en`, you no-op the code loading as a special case. So like for`en_US`:

1. try to load "en-us.js"
2. fails
3. OK, fallback, try to load "en"
4. special case skips loading the code and just does `moment.locale('en')`
5. win

IOW it's "en" you want to special case, not 'en_US`. It's worth noting that Moment's `en` defaults are US; that's why we don't bother with an 'en-us.js' file.

**Comment by codeofsumit** [comment]
@icambron ah I understand. I'll try this next week and will update here accordingly. Thanks for the support!

**Comment by codeofsumit** [comment]
@icambron ok I special cased `en_US` to fall back to regular `en` or `en-gb` - it works but it doesn't seem to be the only special case I need.

The standard language code for japan is `ja_JP`. The filename in moment is `ja.js`.
I think my way to convert lang codes to moment filenames isn't the right way. As far as I can see there is no way around a mapping for each language code to the corresponding filename in moment.js :-(

**Comment by codeofsumit** [comment]
maybe at some point in the future, we could rename the locale filenames to consistently reflect standardized language codes, including en-us, ja-jp or de-de, so we can programmatically find the filename by language code.
